### ISSUE TITLE
Allow different digest_algorithm and signature_algorithm when validating CMS

### ISSUE THREAD

**Issue by peteris-zealid** [MatthiasValvekens/pyHanko#242]
**Is your feature request related to a problem? Please describe.**
Yes, id documents with nfc chips have something called the security object (sod). This sod contains some hashes of personal information and a CMS structure signing those hashes. The signature is produced by a document issuing authority of a given country so there is not much that can be done to influence the signing process.
For Lietuvas documents these digests are in fact different. (I will check other countries in the following days)

**Describe the solution you'd like**
If I understand rfc6211 correctly the requirement for the two algorithms to be identical is not too strict.
```
   digestAlgorithm  has been implicitly protected by the fact that CMS
      has only defined one digest algorithm for each hash value length.
      (The algorithm RIPEMD-160 was never standardized.)  There is also
      an unwritten convention that the same digest algorithm should be
      used both here and for the signature algorithm.  If newer digest
      algorithms are defined so that there are multiple algorithms for a
      given hash length (it is expected that the SHA-3 project will do
      so), or that parameters are defined for a specific algorithm, much
      of the implicit protection will be lost.
```
I would suggest adding a flag (on validation_context) that would allow skipping this particular check.
These are the lines in question:
```
        if (
            sig_hash_algo is not None
            and sig_hash_algo.dump() != hash_algo_obj.dump()
        ):
            raise SignatureValidationError(
```

**Describe alternatives you've considered**
Not much right now.

**Additional context**
I am describing this as a feature not a bug because the assumption that both digest algorithms will match is very reasonable.

Here is base64 of an LT document [sod](https://lapo.it/asn1js/#d4IE-TCCBPUGCSqGSIb3DQEHAqCCBOYwggTiAgEDMQ0wCwYJYIZIAWUDBAIBMIHnBgZngQgBAQGggdwEgdkwgdYCAQAwCwYJYIZIAWUDBAIBMIHDMCUCAQEEIHk11CGlPDKXQimsPu7HsTwY-t31A_r2gGurv0n-M0KzMCUCAQIEIDxhfCMzK2BCOrj2RPs-ypMm5ftmldqKnp-EUGaAuHL7MCUCAQMEIFFxhWMgBZkXJ6LjVra8qsBFmQmDKSEcYrBwHIPVUKq1MCUCAQsEIO5XHb33PX02kJV1FyYoedJytCCaOU79vTftBbY3pqfGMCUCAQ4EIL3Lkv9ozKhGgPgvkYJoRXDNBv_9oT7KlRDcouM5iVFZoIIC6TCCAuUwggKLoAMCAQICAiePMAsGByqGSM49BAEFADBDMQswCQYDVQQGEwJMVDEXMBUGA1UECgwOQURJQyB1bmRlciBNT0kxDTALBgNVBAMMBENTQ0ExDDAKBgNVBAUTAzAwNjAeFw0yMTEyMDMxMTA4MTJaFw0zMjA0MDMxMTA4MTJaMGQxCzAJBgNVBAYTAkxUMRcwFQYDVQQKDA5BRElDIHVuZGVyIE1PSTEuMCwGA1UEAwwlRG9jdW1lbnQgU2lnbmVyIENlcnRpZmljYXRlIExpdGh1YW5pYTEMMAoGA1UEBRMDMDc4MIIBKjCB6wYHKoZIzj0CATCB3wIBATAoBgcqhkjOPQEBAh0A_____________________wAAAAAAAAAAAAAAATBTBBz____________________-_______________-BBy0BQqFDASzq_VBMlZQRLC317_YuicLOUMjVf-0AxUAvXE0R5nVx_zcRbWfo7mrj2qUi8UEOQS3Dgy9a7S_fzITkLlKA8HTVsIRIjQygNYRXB0hvTdjiLX3I_tMIt_mzUN1oFoHR2RE1YGZhQB-NAIdAP__________________FqLguPA-E90pRVxcKj0CAQEDOgAE8dehH4xTSZca9jvtS_fUFNCLhx-lxqqFuSxNH3oLOwqTNpgRd8jw98RgkMBPLA5GX867cSOnNC6jejB4MCsGA1UdEAQkMCKADzIwMjExMjAyMjIwMDAwWoEPMjAyMjAzMDIyMTU5NTlaMBgGA1UdIAQRMA8wDQYLKwYBBAGChlUBAQEwHwYDVR0jBBgwFoAUnSxc6TvhgYf7so4HqIG_PrJjOrgwDgYDVR0PAQH_BAQDAgeAMAsGByqGSM49BAEFAANHADBEAiAJx4vlPIRqZ1p5FNkv7LKCJduy4VP79fJLszapOstClgIgcQ1cMSo4-njHatG2y2QD4ZglrdeLgp7o9kxQuDimXZkxgfYwgfMCAQEwSTBDMQswCQYDVQQGEwJMVDEXMBUGA1UECgwOQURJQyB1bmRlciBNT0kxDTALBgNVBAMMBENTQ0ExDDAKBgNVBAUTAzAwNgICJ48wCwYJYIZIAWUDBAIBoEgwFQYJKoZIhvcNAQkDMQgGBmeBCAEBATAvBgkqhkiG9w0BCQQxIgQgyhrowzQ09bb48wMHgfGl-4KOBA7H52RBtfDJ_ZewQpwwCgYIKoZIzj0EAwEEQDA-Ah0Ah54gj7QxgstG8as64yfSOHZwvIbKkTmRRSx69wIdANZx3JonAujp4_k0m-MpMelexWpBS5KeZNKoqXs)

**Comment by peteris-zealid** [comment]
Also from rfc6211
```
There is a convention, undocumented as far as I
can tell, that the same hash algorithm should be used for both the
content digest and the signature digest.  There are cases, such as
third-party signers that are only given a content digest, where such
a convention cannot be enforced.
```

**Comment by MatthiasValvekens** [comment]
Hi @peteris-zealid,

pyHanko interprets CMS (RFC 5652) as updated in RFC 8933, which makes that convention mandatory. But I can try to add in an override. I'm not yet completely sure where, though, since `ValidationContext` is about x509 validation more than about signature checks, but that's a relatively minor detail.

**Comment by peteris-zealid** [comment]
Yes, you are correct. I now looked at RFC8933. 

An override would be much appreciated, because I do not see how I could hack around this if clause in a way that is reasonably elegant.

With regards to implementation putting the override in `AlgorithmUsagePolicy` would probably make more sense, but it is still part of validation context. Not sure about the semantics of all these things.

In general the internal api of certvalidator is better after the refactor. I did not inspect too much, but at first glance it looks really good. Well done!
